statsmodels/sandbox/examples/gam_gsoc2015/gam.py

```diff
@@ -0,0 +1,267 @@
+                res_aux = self._fit_zeros(keep_index, **kwds)
+                return res_aux
+            else:
+                return res
+
+## this class will be later removed and taken from another push
+class Penalty(object):
+    """
+    A class for representing a scalar-value penalty.
+    Parameters
+    wts : array-like
+        A vector of weights that determines the weight of the penalty
+        for each parameter.
+    Notes
+    -----
+    The class has a member called `alpha` that scales the weights.
+    """
+
+    def __init__(self, wts):
+        self.wts = wts
+        self.alpha = 1.
+
+    def func(self, params):
+        """
+        A penalty function on a vector of parameters.
+        Parameters
+x        ----------
+        params : array-like
+            A vector of parameters.
+        Returns
+        -------
+        A scalar penaty value; greater values imply greater
+        penalization.
+        """
+        raise NotImplementedError
+
+    def grad(self, params):
+        """
+        The gradient of a penalty function.
+        Parameters
+        ----------
+        params : array-like
+            A vector of parameters
+        Returns
+        -------
+        The gradient of the penalty with respect to each element in
+        `params`.
+        """
+        raise NotImplementedError
+
+
+class GamPenalty(Penalty):
+    
+    def __init__(self, wts=1, alpha=1, cov_der2=None, der2=None):
+        
+        self.wts = wts #should we keep wts????
+        self.alpha = alpha
+        self.cov_der2 = cov_der2
+        self.der2 = der2
+    
+    def func(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        '''
+        
+        # The second derivative of the estimated regression function
+        f = np.dot(self.der2, params)
+
+        return self.alpha * np.sum(f**2)
+
+    def grad(self, params):
+        ''' 
+        1) params are the coefficients in the regression model
+        2) der2  is the second derivative of the splines basis
+        3) cov_der2 is obtained as np.dot(der2.T, der2)
+        '''
+                
+        return 2 * self.alpha * np.dot(self.cov_der2, params) 
+    
+    def deriv2(self, params):
+
+        return 2 * self.alpha * self.cov_der2
+
+
+
+class LogitGam(PenalizedMixin, Logit):
+    pass
+
+
+class GLMGam(PenalizedMixin, GLM):
+    pass
+  
+
+
+def test_gam_penalty():
+    ''' test the gam penalty class '''
+    n = 100000
+    x = np.linspace(-10, 10, n)
+    degree = 3
+    basis, der_basis, der2_basis = make_poly_basis(x, degree)
+    cov_der2 = np.dot(der2_basis.T, der2_basis)
+    gp = GamPenalty(alpha=1, der2=der2_basis, cov_der2=cov_der2)
+    params = np.array([1, 1, 1, 1])
+    cost = gp.func(params) 
+    # the integral between -10 and 10 of |2*a+6*b*x|^2 is 80*a^2 + 24000*b^2
+    assert(int(cost/n*20) == 24080)
```

Comment:
in general, we use from `numpy.testing import assert_allclose` and other related asserts
No plain python asserts